app/code/Magento/CatalogUrlRewrite/Observer/ProductProcessUrlRewriteSavingObserver.php

```diff
@@ -3,73 +3,167 @@
+    private $storeWebsiteRelation;
+
+    /**
+     * @var AppendUrlRewritesToProducts
+     */
+    private $appendRewrites;
 
     /**
-     * @var ProductUrlPathGenerator
+     * @var ScopeConfigInterface
      */
-    private $productUrlPathGenerator;
+    private $scopeConfig;
 
     /**
-     * @param ProductUrlRewriteGenerator $productUrlRewriteGenerator
      * @param UrlPersistInterface $urlPersist
-     * @param ProductUrlPathGenerator|null $productUrlPathGenerator
+     * @param StoreWebsiteRelationInterface|null $storeWebsiteRelation
+     * @param AppendUrlRewritesToProducts|null $appendRewrites
+     * @param ScopeConfigInterface $scopeConfig
      */
     public function __construct(
-        ProductUrlRewriteGenerator $productUrlRewriteGenerator,
         UrlPersistInterface $urlPersist,
-        ProductUrlPathGenerator $productUrlPathGenerator = null
+        StoreWebsiteRelationInterface $storeWebsiteRelation,
+        AppendUrlRewritesToProducts $appendRewrites,
+        ScopeConfigInterface $scopeConfig
     ) {
-        $this->productUrlRewriteGenerator = $productUrlRewriteGenerator;
         $this->urlPersist = $urlPersist;
-        $this->productUrlPathGenerator = $productUrlPathGenerator ?: ObjectManager::getInstance()
-            ->get(ProductUrlPathGenerator::class);
+        $this->storeWebsiteRelation = $storeWebsiteRelation;
+        $this->appendRewrites = $appendRewrites;
+        $this->scopeConfig = $scopeConfig;
     }
 
     /**
      * Generate urls for UrlRewrite and save it in storage
      *
-     * @param \Magento\Framework\Event\Observer $observer
+     * @param Observer $observer
      * @return void
-     * @throws \Magento\UrlRewrite\Model\Exception\UrlAlreadyExistsException
      */
-    public function execute(\Magento\Framework\Event\Observer $observer)
+    public function execute(Observer $observer)
     {
         /** @var Product $product */
         $product = $observer->getEvent()->getProduct();
 
-        if ($product->dataHasChangedFor('url_key')
-            || $product->getIsChangedCategories()
-            || $product->getIsChangedWebsites()
-            || $product->dataHasChangedFor('visibility')
-        ) {
-            if ($product->isVisibleInSiteVisibility()) {
-                $product->unsUrlPath();
-                $product->setUrlPath($this->productUrlPathGenerator->getUrlPath($product));
-                $this->urlPersist->replace($this->productUrlRewriteGenerator->generate($product));
+        if ($this->isNeedUpdateRewrites($product)) {
+            $this->deleteObsoleteRewrites($product);
+            $oldWebsiteIds = $product->getOrigData('website_ids');
+            $storesToAdd = [];
+            if ($oldWebsiteIds !== null) {
+                $storesToAdd = $this->getStoresListByWebsiteIds(
+                    array_diff($product->getWebsiteIds(), $oldWebsiteIds)
+                );
+            }
+            $this->appendRewrites->execute([$product], $storesToAdd);
```

Comment:
Don't need to handle this exception. This is expected behavior in case when admin tries to save a product with url-key that already assigned to another item either product or category. See \Magento\CatalogUrlRewrite\Model\ProductUrlRewriteTest::testUrlRewriteOnProductSaveWithExistingUrlKey

Reply:
In this case please add exception to annotaions.

Reply:
added